components/feature/sendtab/src/main/java/mozilla/components/feature/sendtab/SendTabFeature.kt

```diff
@@ -0,0 +1,150 @@
+    accountManager: FxaAccountManager,
+    pushFeature: AutoPushFeature? = null,
+    pushService: PushService? = null,
+    owner: LifecycleOwner = ProcessLifecycleOwner.get(),
+    autoPause: Boolean = false,
+    onTabsReceived: (Device?, List<TabData>) -> Unit
+) {
+    init {
+        val accountObserver = AccountObserver(context, pushFeature, pushService)
+        val pushObserver = PushObserver(accountManager)
+        val deviceObserver = DeviceObserver(onTabsReceived)
+
+        // Always observe the account for device events.
+        accountManager.registerForDeviceEvents(deviceObserver, owner, autoPause)
+
+        pushFeature?.apply {
+            registerForPushMessages(PushType.Services, pushObserver, owner, autoPause)
+            registerForSubscriptions(pushObserver, owner, autoPause)
+
+            // observe the account only if we have the push feature (service is optional)
+            accountManager.register(accountObserver, owner, autoPause)
+        }
+    }
+}
+
+internal class PushObserver(
+    private val accountManager: FxaAccountManager
+) : Bus.Observer<PushType, String>, PushSubscriptionObserver {
+    private val logger = Logger("PushObserver")
+
+    override fun onSubscriptionAvailable(subscription: AutoPushSubscription) {
+        logger.debug("Received new push subscription from $subscription.type")
+
+        if (subscription.type == PushType.Services) {
+            accountManager.withConstellation {
+                it.setDevicePushSubscriptionAsync(
+                    DevicePushSubscription(
+                        endpoint = subscription.endpoint,
+                        publicKey = subscription.publicKey,
+                        authKey = subscription.authKey
+                    )
+                )
+            }
+        }
+    }
+
+    override fun onEvent(type: PushType, message: String) {
+        logger.debug("Received new push message for $type")
+
+        accountManager.withConstellation {
+            it.processRawEventAsync(message)
+        }
+    }
+}
+
+internal class DeviceObserver(
+    private val onTabsReceived: (Device?, List<TabData>) -> Unit
+) : DeviceEventsObserver {
+    private val logger = Logger("DeviceObserver")
+
+    override fun onEvents(events: List<DeviceEvent>) {
+        events.asSequence()
+            .filterIsInstance<DeviceEvent.TabReceived>()
+            .forEach { event ->
+                logger.debug("Showing ${event.entries.size} tab(s) received from deviceID=${event.from?.id}")
+
+                onTabsReceived(event.from, event.entries)
+            }
+    }
+}
+
+internal class AccountObserver(
+    private val context: Context,
+    private val feature: AutoPushFeature?,
+    private val service: PushService?
+) : SyncAccountObserver {
+    private val logger = Logger("AccountObserver")
+
+    override fun onAuthenticated(account: OAuthAccount, newAccount: Boolean) {
+        service?.start(context)
+
+        if (newAccount) {
```

Comment:
Can you add a comment describing why we only subscribe once per account lifecycle?